File: apps/mobile/android/app/src/main/kotlin/com/cyberremesher/mobile/ui/UiModel.kt
// UNVERIFIED: Android shell — decoders for the shared data-driven UI model
// (task 8.6). Not built in CI. Mirrors ../../shared/stages.json and
// toolbar.default.json (same files the iPadOS shell decodes). Loaded from the
// app's assets/ (copied there by the Gradle `copySharedUiModel` task).

package com.cyberremesher.mobile.ui

import android.content.Context
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class Stage(
    val id: String,
    val title: String,
    @SerialName("androidIcon") val androidIcon: String = "",
    val summary: String = "",
)

@Serializable
data class ActionDescriptor(
    val id: String,
    val title: String,
    @SerialName("androidIcon") val androidIcon: String = "",
    val stages: List<String> = emptyList(),
    val command: String,
) {
    /** True if this action is available in [stageId] (empty stages = all). */
    fun availableIn(stageId: String): Boolean =
        stages.isEmpty() || stages.contains(stageId)
}

@Serializable
data class StageModel(val schemaVersion: Int = 1, val stages: List<Stage> = emptyList())

@Serializable
data class ToolbarModel(
    val schemaVersion: Int = 1,
    val actions: List<ActionDescriptor> = emptyList(),
    val defaultToolbar: List<String> = emptyList(),
) {
    /** The default toolbar resolved to descriptors, in order. */
    fun defaultActions(): List<ActionDescriptor> {
        val byId = actions.associateBy { it.id }
        return defaultToolbar.mapNotNull { byId[it] }
    }
}

/** Loads the shared UI model from assets, falling back to a minimal built-in. */
object UiModelLoader {
    private val json = Json { ignoreUnknownKeys = true }

    fun loadStages(context: Context): StageModel =
        decode(context, "stages.json") ?: StageModel(
            stages = listOf(
                Stage("retopology", "Retopology"),
                Stage("uv", "UV"),
                Stage("baking", "Baking"),
            )
        )

    fun loadToolbar(context: Context): ToolbarModel =
        decode(context, "toolbar.default.json") ?: ToolbarModel()

    private inline fun <reified T> decode(context: Context, asset: String): T? =
        runCatching {
            context.assets.open(asset).bufferedReader().use { reader ->
                json.decodeFromString<T>(reader.readText())
            }
        }.getOrNull()
}

File: apps/mobile/ipados/Sources/CyberRemesherApp/UIModel.swift
// UNVERIFIED: iPadOS shell — decoders for the shared data-driven UI model
// (task 8.6). Pure Swift + Foundation (no UIKit), so it type-checks with
// `swift build`. Mirrors `../../shared/stages.json` and
// `../../shared/toolbar.default.json`; the Android shell decodes the same files.

import Foundation

/// One editing stage in the stage switcher (Retopology → UV → Baking).
public struct Stage: Codable, Identifiable, Equatable, Sendable {
    public let id: String
    public let title: String
    /// SF Symbol name for the iPadOS tab; the Android shell reads `androidIcon`.
    public let systemImage: String
    public let summary: String
}

/// One action (tool) that can appear in the toolbar or the Action Gallery.
public struct ActionDescriptor: Codable, Identifiable, Equatable, Sendable {
    public let id: String
    public let title: String
    public let systemImage: String
    /// Stage ids in which the action is meaningful; empty = all stages.
    public let stages: [String]
    /// Opaque engine chord/command id the shell injects (routing is the
    /// engine's concern — the shell never interprets it).
    public let command: String
}

/// Decoded `stages.json`.
public struct StageModel: Codable, Sendable {
    public let schemaVersion: Int
    public let stages: [Stage]
}

/// Decoded `toolbar.default.json`.
public struct ToolbarModel: Codable, Sendable {
    public let schemaVersion: Int
    public let actions: [ActionDescriptor]
    /// Ordered action ids shown in the toolbar by default.
    public let defaultToolbar: [String]
}

/// Loads the shared UI model from the app bundle, falling back to a minimal
/// built-in so the shell still renders if the resources are missing.
public enum UIModelLoader {
    public static func loadStages(bundle: Bundle = .main) -> StageModel {
        decode("stages", as: StageModel.self, bundle: bundle)
            ?? StageModel(schemaVersion: 1, stages: [
                Stage(id: "retopology", title: "Retopology",
                      systemImage: "scribble.variable", summary: ""),
                Stage(id: "uv", title: "UV", systemImage: "map", summary: ""),
                Stage(id: "baking", title: "Baking", systemImage: "paintbrush",
                      summary: ""),
            ])
    }

    public static func loadToolbar(bundle: Bundle = .main) -> ToolbarModel {
        decode("toolbar.default", as: ToolbarModel.self, bundle: bundle)
            ?? ToolbarModel(schemaVersion: 1, actions: [], defaultToolbar: [])
    }

    private static func decode<T: Decodable>(
        _ resource: String, as _: T.Type, bundle: Bundle
    ) -> T? {
        guard let url = bundle.url(forResource: resource, withExtension: "json"),
              let data = try? Data(contentsOf: url)
        else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }
}
